amaranth/lib/io.py

```diff
@@ -253,6 +258,362 @@ def __repr__(self):
+        if not (hasattr(port, "direction") and isinstance(port.direction, Direction)):
+            raise TypeError(f"'port' must be a 'SingleEndedPort', 'DifferentialPort', "
+                            f"or a platform-specific object, not {port!r}")
+        self._port = port
+        super().__init__(Buffer.Signature(direction, len(port)))
+        if port.direction is Direction.Input and self.direction is not Direction.Input:
+            raise ValueError(f"Input port cannot be used with {self.direction.name} buffer")
+        if port.direction is Direction.Output and self.direction is not Direction.Output:
+            raise ValueError(f"Output port cannot be used with {self.direction.name} buffer")
+
+    @property
+    def port(self):
+        return self._port
+
+    @property
+    def direction(self):
+        return self.signature.direction
+
+    def elaborate(self, platform):
+        if hasattr(platform, "get_io_buffer"):
+            res = platform.get_io_buffer(self)
+            if res is not None:
+                return res
+
+        m = Module()
+
+        invert = sum(bit << idx for idx, bit in enumerate(self._port.invert))
+        if self.direction is not Direction.Input:
+            if invert:
+                o_inv = Signal.like(self.o)
+                m.d.comb += o_inv.eq(self.o ^ invert)
+            else:
+                o_inv = self.o
+        if self.direction is not Direction.Output:
+            if invert:
+                i_inv = Signal.like(self.i)
+                m.d.comb += self.i.eq(i_inv ^ invert)
+            else:
+                i_inv = self.i
+
+        if isinstance(self._port, SingleEndedPort):
+            if self.direction is Direction.Input:
+                m.submodules.iob = IOBufferInstance(self._port.io, i=i_inv)
+            elif self.direction is Direction.Output:
+                m.submodules.iob = IOBufferInstance(self._port.io, o=o_inv, oe=self.oe)
+            else:
+                m.submodules.iob = IOBufferInstance(self._port.io, o=o_inv, oe=self.oe, i=i_inv)
+        elif isinstance(self._port, DifferentialPort):
+            if self.direction is Direction.Input:
+                m.submodules.iob_p = IOBufferInstance(self._port.p, i=i_inv)
+            elif self.direction is Direction.Output:
+                m.submodules.iob_p = IOBufferInstance(self._port.p, o=o_inv, oe=self.oe)
+                m.submodules.iob_n = IOBufferInstance(self._port.n, o=~o_inv, oe=self.oe)
+            else:
+                m.submodules.iob_p = IOBufferInstance(self._port.p, o=o_inv, oe=self.oe, i=i_inv)
+                m.submodules.iob_n = IOBufferInstance(self._port.n, o=~o_inv, oe=self.oe)
+        else:
+            raise TypeError("Cannot elaborate generic 'Buffer' with port {self._port!r}")
+
+        return m
+
+
+class FFBuffer(wiring.Component):
+    """A registered I/O buffer.
+
+    Equivalent to a plain :class:`Buffer` combined with reset-less registers on ``i``, ``o``, ``oe``.
```

Comment:
These should be using `:py:`. (Anything making reference to Python tokens should be using that.)